sql/schemachanger:  enable create index by default

fqazi opened the pull request:
Epic: CRDB-19164

This patch will enable CREATE INDEX by default inside the declarative schema changer by fixing the following:
1. Fix validation for supported types for inverted indexes
2. Implement full inverted index support
3. Re-use and properly create columns for index expressions
4. Add support for partial indexes
5. Add support for partitioned indexes
6. Address some hash-shared index bugs, related to wrong column types. Support is still disabled by default since we need constraint support

fqazi committed:
sql/schemchanger: disable declarative schema changer in incompatible tests

Some tests test low level functionality of CREATE INDEX legacy
schema changer hooks. We will temporarily disable declarative
schema changer in these tests. Additionally, update logic
test files for new expected output from declarative schema
changer.

Release note: None

fqazi committed:
sql/schemachanger: adopt index spec for create index

Previously, the CREATE INDEX code in the declarative schema
changer manually added elements. We can retractor the code
to be a bit cleaner by using the indexSpec object

Release note: None

fqazi committed:
sql/schemachanger: generate index names for secondary indexes

Previously, we lacked logic to generate secondary index names,
when no name was specified. This patch adds the required logic.

Release note: None

fqazi committed:
sql/schemachanger: create index add validation for types and storing columns

Previously, the declarative schema changer code did not bother to validate
if any of the types used in an index were indexable. Addtionally, there were
bugs in our duplicate column detection for stored values.

Release note: None:

fqazi committed:
sql/schemachanger: create index in declarative schema changer enable expression reuse

Previously, the declarative schema changer supported creating indexes
with expressions with no support for reusing virtual columns. Additionally,
there were other bugs in this expression logic likes accessibility flags
and lack of type checking. This patch addresses all of those concerns.

Release note: None

fqazi committed:
sql/schemachanger: add implicit columns and inverted index kind into decomposition

To help support partitioning the decomposition needs to be able to
track the kind of inverted index columns (trigram vs default) and
which columns are implicitly there in an index.

Release note: None

fqazi committed:
sql/schemachanger: enable inverted index support for non-geo types

Previously, inverted index was not properly supported for non-geo
types and we would fall back to the legacy schema changer. This patch
enables it for the declarative schema changer when non-geo types are
used.

Release note: None

fqazi committed:
sql/scheamchanger: add support for partitioning inside CREATE INDEX

Previously, the declarative schema changer did not support
CREATE INDEX statements with partitions. This path adds
support for them.

Release note: None

fqazi committed:
sql/schemachanger: add support for geo types in indexes

Previously, the declarative schema changer did not support
geometry/geographic types in CREATE INDEX. This patch
adds support for them including geoconfig parameters
during CREATE INDEX.

Release note: None

fqazi committed:
sql/schemachanger: fix hash sharded index bugs

Previously, our hash sharded index implementations had bugs,
which would prevent us from enabling CREATE INDEX by default.
This patch disables it support for CREATE INDEX in the declarative
schema changer if hash sharded indexes are detected. It also fixes
the types for the columns used, so that it behaves correctly,
once other pieces are in place to support it.

Release note: None

fqazi committed:
sql/schemachanger: fix execution bugs for create index

Previously the declarative schema changer did not correctly
do the following for CREATE INDEX:
1) Correctly return execution errors from validation.
2) Generate event log entries for CREATE INDEX.
3) Ensure that index names are assigned before validation, so
   proper errors are returned.
4) Enable CREATE INDEX support by default.

Release note: None

fqazi committed:
sql/schemachanger: enable create index inside declarative by default

Previously, CREATE INDEX ran by default inside the legacy schema changer,
these changes enable it inside the declarative one.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/92128)
<!-- Reviewable:end -->

fqazi committed:
sql/schemachanger: add telemetry for CREATE INDEX

Previously, the CREATE INDEX logic inside the declarative
schema changer did not implement telemetry. This patch
adds telemetry counters for different variants of
the CREATE INDEX statement.

Release note: None

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -212,6 +215,131 @@ func nextRelationIndexID(b BuildCtx, relation scpb.Element) catid.IndexID {
+		}
+	}
+}
+
```
nit: similarly, can you some comment to explain what this function does in a high level?

Xiang-Gu replied:
Question: Is this logic copied/mimiced from somewhere else?

chengxiong-ruan replied:
it should be :) it's mostly from `create_index.go` of legacy schema changer, for example: https://github.com/cockroachdb/cockroach/blob/b6e661e4ff03bd7c9650c703f1fb98dc213cad63/pkg/sql/create_index.go#L330-L330

fqazi replied:
Added the comment:

// checkColumnAccessibilityForIndex validate that any columns that are explicitly referenced in a column for storage or as a key are either accessible and not system columns.

fqazi replied:
> Question: Is this logic copied/mimiced from somewhere else?

Yes, this is taken from, checkIndexColumns

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -212,6 +215,131 @@ func nextRelationIndexID(b BuildCtx, relation scpb.Element) catid.IndexID {
 	}
 }
 
+func newUndefinedOpclassError(opclass tree.Name) error {
```
Question: what is an operator class?

chengxiong-ruan replied:
It's mostly for inverted indexes: https://www.postgresql.org/docs/current/gin-builtin-opclasses.html
But looks like only geometry and trigram are taken care of in crdb: https://github.com/cockroachdb/cockroach/blob/b6e661e4ff03bd7c9650c703f1fb98dc213cad63/pkg/sql/create_index.go#L367-L367

fqazi replied:
@Xiang-Gu https://www.postgresql.org/docs/current/indexes-opclass.html, it pretty much determines how values are compared / stored for a given index. I'll make add comments in the areas where this is checked:

// OpClass in Cockroach is used to determine how values will be compared/stored within the index,
// similar to Postgres.

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -283,6 +414,20 @@ func addColumnsForSecondaryIndex(
 			e.Kind != scpb.IndexColumn_KEY {
 			return
 		}
+		// Check if the column name was duplicated from the STORING clause, in which
+		// case this isn't allowed.
+		// Note: The column IDs for the key suffix columns are not resolved, so we couldn't
+		// do this earlier.
+		scpb.ForEachColumnName(relationElements, func(current scpb.Status, target scpb.TargetStatus, cn *scpb.ColumnName) {
+			if cn.ColumnID == e.ColumnID &&
+				cn.TableID == e.TableID {
+				if _, found := columnRefs[cn.Name]; found {
+					panic(pgerror.Newf(pgcode.InvalidObjectDefinition,
+						"index %q already contains column %q", n.Name, cn.Name))
+				}
+				columnRefs[cn.Name] = struct{}{}
+			}
+		})
```
1. Why do we have to check "whether there is duplicate STORING columns" here, but not earlier? I didn't understand the comment "Note: The column IDs for the key suffix columns are not resolved, so we couldn't do this earlier."

2. You are computing the name of column `e`, right? If so, how is this related to STORING columns in the to-be-added index as `e` is a key suffix column in the to-be-added index?

3. I think you can just call function `mustRetrieveColumnNameElem` to get the name of a column by specifying table ID and column ID.

fqazi replied:
1) The key suffix columns are derived from the primary key, pretty much any column that isn't contained by the key columns in the index. To make this clearer I'll change this comment to:
// Note: The column IDs for the key suffix columns are derived by finding columns in the primary index, which
// are not covered by the key columns within the index. We checked against the key columns vs the storing ones,
// earlier so this covers any extra columns.

2) If the column is already in the key either implicitly via the suffix or directly by the references key columns, then it makes no sense for it to be stored again. So, this check is pretty much saying the set of storing columns and key suffix columns don't overlap.

3) TIL, let me replace of these

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -536,11 +546,82 @@ func maybeCreateAndAddShardCol(
 
-func createVirtualColumnForIndex(
-	b BuildCtx, tn *tree.TableName, tbl *scpb.Table, expr tree.Expr,
+func maybeCreateVirtualColumnForIndex(
+	b BuildCtx, tn *tree.TableName, tbl *scpb.Table, expr tree.Expr, inverted bool, lastColumn bool,
```
TIL: virtual columns can be used in more than one index!

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -536,11 +546,82 @@ func maybeCreateAndAddShardCol(
 	elts := b.QueryByID(tbl.TableID)
-	colName := tabledesc.GenerateUniqueName("crdb_internal_idx_expr", func(name string) (found bool) {
+	colName := ""
+	// Check if any existing columns can satisfy this expression already.
```
IIUC, you are looking for a column `e` (with ToPublic target) who has an compute expression that is equal to the compute expression of the to-be-added virtual column. If succeeded, we return early and hence re-using virtual column, right?

But where did you check this existing column `e` is also a virtual column? What if it's stored? Maybe we need the guardrails to be
```if target == scpb.ToPublic && e.ComputeExpr != nil && e.IsVirtual```

fqazi replied:
Good point this should be e.IsVirtual && e.Inaccesible, so that user columns are never used.

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -590,7 +670,16 @@ func createVirtualColumnForIndex(
+		validateColumnIndexableType(typedExpr.ResolvedType())
 	}
 	alterTableAddColumn(b, tn, tbl, &tree.AlterTableAddColumn{ColumnDef: d})
+	// Mutate the accessibility flag on this column it should be inaccessible.
+	{
+		ers := b.ResolveColumn(tbl.TableID, d.Name, ResolveParams{
+			RequiredPrivilege: privilege.CREATE,
+		})
+		_, _, col := scpb.FindColumn(ers)
+		col.IsInaccessible = true
+	}
```
Why? Do you mean a virtual column is always inaccessible?

fqazi replied:
Yes, any new virtual columns added by CREATE INDEX need to be inaccessible, this is the current behaviour. Just being hidden isn't sufficient.

fqazi replied:
I'll add a comment to make it more clear:

// When a virtual column for an index expression gets added for CREATE INDEX it must be inaccessible,
// so we will manipulate the newly added element to be inaccessible after going through the normal add
// column code path.

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/builder_state.go:
```diff
@@ -536,6 +536,25 @@ func (b *builderState) ComputedColumnExpression(tbl *scpb.Table, d *tree.ColumnT
+) *scpb.Expression {
+	_, _, ns := scpb.FindNamespace(b.QueryByID(tableID))
+	tn := tree.MakeTableNameFromPrefix(b.NamePrefix(ns), tree.Name(ns.Name))
+	// TODO(fqazi): this doesn't work when referencing newly added columns.
+	_, err := schemaexpr.ValidatePartialIndexPredicate(
+		b.ctx,
+		b.descCache[tableID].desc.(catalog.TableDescriptor),
+		expr,
+		&tn,
+		b.semaCtx)
+	if err != nil {
+		panic(err)
+	}
```
I encountered exactly the same issue when implementing `add constraint check` where I also need to validate the check expression in the builder. What I did is to refactor the code  in `schemaexpr.DequalifyAndValidateExpr` so that the caller will need to provide the columns, so I can circumvent this limitation.

You will benefit from that refactoring as well but this PR does not need to be blocked by it.

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_index_geoconfig.go:
```diff
@@ -0,0 +1 @@
+package opgen
```
oh I see it now -- it's for use in the next commit. Please fix it.

Xiang-Gu commented on pkg/sql/schemachanger/scpb/elements.proto:
```diff
@@ -95,6 +96,7 @@ message ElementProto {
   IndexComment index_comment = 43 [(gogoproto.moretags) = "parent:\"PrimaryIndex, SecondaryIndex\""];
   IndexColumn index_column = 44 [(gogoproto.moretags) = "parent:\"PrimaryIndex, SecondaryIndex, TemporaryIndex, Column\""];
   IndexData index_data = 45 [(gogoproto.customname) = "IndexData", (gogoproto.moretags) = "parent:\"PrimaryIndex, SecondaryIndex, TemporaryIndex\""];
+  IndexGeoConfig index_geo_config = 46  [(gogoproto.moretags) = "parent:\"PrimaryIndex, SecondaryIndex, TemporaryIndex\""];
```
I am curious what is the justification/benefit for introducing new elements for `geo config` and `secondary_index_partial`, instead of adding them as fields in existing elements?

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_secondary_index.go:
```diff
@@ -86,6 +86,14 @@ func init() {
 				}),
 			),
 			to(scpb.Status_PUBLIC,
+				emit(func(this *scpb.SecondaryIndex) *scop.RefreshStats {
+					return &scop.RefreshStats{
+						TableID: this.TableID,
+					}
+				}),
+				emit(func(this *scpb.SecondaryIndex, md *opGenContext) *scop.LogEvent {
+					return newLogEventOp(this, md)
+				}),
```
I am curious is there a reason that we need to emit those two operations before `MakeValidatedSecondaryIndexPublic`? Or it does not matter?

fqazi replied:
It honestly doesn't matter in this case, since if the later operation fails none of these will do anything either. These commands only have a side effect once a stage has finished execution.

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/rules/dep_add_index.go:
```diff
@@ -71,21 +71,36 @@ func init() {
+					(*scpb.SecondaryIndex)(nil),
+				),
+				joinOnIndexID(from, to, "table-id", "index-id"),
+				statusesToPublicOrTransient(from, scpb.Status_PUBLIC, to, scpb.Status_WRITE_ONLY),
```
1. What is the overall rational that prompts you to make this dep-rule change?
2. Specifically for this rule, why SameStagePrecedence? If your answer is "It previously was SameStagePrecedence, and I just carried it over", then I am starting to think we can change both to "Precedence"? What do you think?

fqazi replied:
1. When we validate an index any validation error will contain the name of the index, so we need to set the index name before validation occurs. 
2. We can safely make this precedence since validation ops will have their own independent stage no matter what, so I'm okay with that change.

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go:
```diff
@@ -69,7 +69,7 @@ var supportedStatements = map[reflect.Type]supportedStatement{
 	// here.
 	reflect.TypeOf((*tree.AlterTable)(nil)):          {fn: AlterTable, on: true, extraChecks: alterTableIsSupported},
-	reflect.TypeOf((*tree.CreateIndex)(nil)):         {fn: CreateIndex, on: false, minSupportedClusterVersion: clusterversion.V22_2Start},
+	reflect.TypeOf((*tree.CreateIndex)(nil)):         {fn: CreateIndex, on: true, minSupportedClusterVersion: clusterversion.V23_1Start},
```
Such a monumental change! Bravo!

Xiang-Gu commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -8339,6 +8339,10 @@ func TestBackupOnlyPublicIndexes(t *testing.T) {
 
+	// Disable declarative schema changer for this test, until the knobs are updated.
+	sqlDB.Exec(t, "SET use_declarative_schema_changer='off';")
+	sqlDB.Exec(t, "SET CLUSTER SETTING sql.defaults.use_declarative_schema_changer='off';")
```
Educational: This is the first time I see this cluster setting. I've always known and only know the session setting. What is this cluster setting and why do we have it?

fqazi replied:
It sets the default setting for the session variable so that any new connection will inherit the same value. This test uses multiple connections, so instead of setting this as disabled for each one, we can disable it across all of them.

Xiang-Gu commented on pkg/sql/sem/tree/constant.go:
```diff
@@ -71,6 +71,12 @@ func typeCheckConstant(
 	if !desired.IsAmbiguous() {
 		for _, typ := range avail {
 			if desired.Equivalent(typ) {
+				if desired.UserDefined() && !desired.IsHydrated() {
+					desired, err = semaCtx.TypeResolver.ResolveTypeByOID(ctx, desired.Oid())
+					if err != nil {
+						return nil, err
+					}
+				}
```
Is this fix required for CREATE INDEX? If so, then I feel like this commit should probably go somewhere before the last two commits? I was just thinking the last two commits should be "Enable CREATE INDEX by default" and "Disable CREATE INDEX in certain tests".

fqazi replied:
Yes, let me re-order the commits this is needed for the partial index expressions, which are running into a bug where things don't get fully hydrated.

postamar replied:
I'm highly suspicious of this change. For a start, there's no guarantee that semaCtx.TypeResolver is not nil. Also, why isn't this new logic pushed down into the corresponding ResolveAsType implementation instead of being added here?

ajwerner reviewed (commented):
Just scrolled through, will give it a closer pass later. 

> Reviewed 1 of 5 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-:-NHQq9Kn8C93vgDeN4qo:bz6jl6e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @msbutler, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 362 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHQnOyT2z4SdrhbFH72:-NHQnOyT2z4SdrhbFH73:b-waun9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/fecb571af812c110dd7e79ef674bc90a20c1f9df/pkg/sql/schemachanger/scbuild/builder_state.go#L362)):*
> ```Go
> 	for _, rangePart := range ret.Range {
> 		if _, ok := partitionNames[rangePart.Name]; ok {
> 			panic(errors.Newf("PARTITION %s: name must be unique (used twice in index %q)",
> ```

Is this an assertion failure or is this an expected error? If the latter, can you give it a pgcode and make sure you can hit it in a logictest?
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 369 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHQnVyuEwdkdRMYcRgj:-NHQnVyv31ob6f9l8QmJ:b-snc3e5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fecb571af812c110dd7e79ef674bc90a20c1f9df/pkg/sql/schemachanger/scbuild/builder_state.go#L369)):*
> ```Go
> 	for _, listPart := range ret.List {
> 		if _, ok := partitionNames[listPart.Name]; ok {
> 			panic(errors.Newf("PARTITION %s: name must be unique (used twice in index %q)",
> ```

same error question
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 545 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHQoxTJCmvooP8xSOCz:-NHQoxTJCmvooP8xSOD-:b-6442xv) ([raw file](https://github.com/cockroachdb/cockroach/blob/fecb571af812c110dd7e79ef674bc90a20c1f9df/pkg/sql/schemachanger/scbuild/builder_state.go#L545)):*
> ```Go
> 	_, _, ns := scpb.FindNamespace(b.QueryByID(tableID))
> 	tn := tree.MakeTableNameFromPrefix(b.NamePrefix(ns), tree.Name(ns.Name))
> 	// TODO(fqazi): this doesn't work when referencing newly added columns.
> ```

What happens? Can we detect this error? 


<!-- Sent from Reviewable.io -->

fqazi commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -350,13 +351,23 @@ func addColumnsForSecondaryIndex(
-		if _, found := columnRefs[colName]; found {
-			panic(pgerror.Newf(pgcode.InvalidObjectDefinition,
-				"index %q contains duplicate column %q", n.Name, colName))
+		if columnNode.Expr == nil {
+			colName := columnNode.Column.Normalize()
+			if _, found := columnRefs[colName]; found {
+				panic(pgerror.Newf(pgcode.InvalidObjectDefinition,
+					"index %q contains duplicate column %q", n.Name, colName))
+			}
+			columnRefs[colName] = struct{}{}
+		} else {
+			colExpr := columnNode.Expr.String()
+			if _, found := columnExprRefs[colExpr]; found {
+				panic(pgerror.Newf(pgcode.InvalidObjectDefinition,
+					"index %q contains duplicate expression", n.Name))
+			}
+			columnExprRefs[colExpr] = struct{}{}
```
Yes, those are the two cases, pretty much either the same column referenced twice or the same expression twice since neither makes sense. I'll add a comment:

// Detect if either the same column or expression repeats twice in this
// index definition. ie.:
// 1) CREATE INDEX idx ON t(i, i) or
// 2) CREATE INDEX idx ON t(lower(i), j, lower(i)).

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92128)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @msbutler, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 554 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqMCKAOSi3Ts4wY-r8-554:-NHV2YBi5U2bMVt8CsFM:bt5liq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e881c871349150bf1ee7ec14e0cc6e9ddba0f2f9/pkg/sql/schemachanger/scbuild/builder_state.go#L554)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

I encountered exactly the same issue when implementing `add constraint check` where I also need to validate the check expression in the builder. What I did is to refactor the code  in `schemaexpr.DequalifyAndValidateExpr` so that the caller will need to provide the columns, so I can circumvent this limitation.

You will benefit from that refactoring as well but this PR does not need to be blocked by it.
</blockquote></details>

Definitely, though that won't be perfect. We will still have issues with type references where if one gets added within a transaction I think.
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 362 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHQnOyT2z4SdrhbFH72:-NHV25yv3Q7fkLCpkl7r:b1mz6ir) ([raw file](https://github.com/cockroachdb/cockroach/blob/fecb571af812c110dd7e79ef674bc90a20c1f9df/pkg/sql/schemachanger/scbuild/builder_state.go#L362)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Is this an assertion failure or is this an expected error? If the latter, can you give it a pgcode and make sure you can hit it in a logictest?
</blockquote></details>

Sadly its an assertion that we test in logic tests, so to preserved behaviour we are validating it here :facepalm: 
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 369 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHQnVyuEwdkdRMYcRgj:-NHV2D0k-nj7bFY0Hbm1:b-djtf9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/fecb571af812c110dd7e79ef674bc90a20c1f9df/pkg/sql/schemachanger/scbuild/builder_state.go#L369)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

same error question
</blockquote></details>

Done.

Same answer.
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 545 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHQoxTJCmvooP8xSOCz:-NHV2KLLCdR9fNkzHp8Q:br2f3fm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fecb571af812c110dd7e79ef674bc90a20c1f9df/pkg/sql/schemachanger/scbuild/builder_state.go#L545)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

What happens? Can we detect this error? 
</blockquote></details>

This is only a problem with transactional support where we add a column and then go reference it here. So, the declarative one in the current implicit mode this is a non-issue.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92128)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @msbutler, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 362 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHQnOyT2z4SdrhbFH72:-NHV7vnP2UWP1uPxzm2t:btqta2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/fecb571af812c110dd7e79ef674bc90a20c1f9df/pkg/sql/schemachanger/scbuild/builder_state.go#L362)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Sadly its an assertion that we test in logic tests, so to preserved behaviour we are validating it here :facepalm: 
</blockquote></details>

Well, it seems like it is just missing a pgcode and we should add one in both places. 


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92128)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @msbutler, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 362 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHQnOyT2z4SdrhbFH72:-NHVAteZEk6P_slm0jsG:brcz0zh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fecb571af812c110dd7e79ef674bc90a20c1f9df/pkg/sql/schemachanger/scbuild/builder_state.go#L362)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Well, it seems like it is just missing a pgcode and we should add one in both places. 
</blockquote></details>

Yes, let me update the existing legacy and declarative code to just assign a proper pgcode for this error.


<!-- Sent from Reviewable.io -->

fqazi committed:
sql/schemachanger: add support for partial indexes

Previously, the CREATE INDEX statement in declarative
schema changer did not support partial indexes. This
patch adds support for them.

Release note: None

fqazi committed:
sql: add pgcode for CREATE INDEX partition definition errors

Previously, invalid partition configurations would
generate generic unexpected errors. This patch, properly
maps these errors to the InvalidObjectDefinition pgcode.

Release note: None

fqazi commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -212,6 +215,131 @@ func nextRelationIndexID(b BuildCtx, relation scpb.Element) catid.IndexID {
+func newUndefinedOpclassError(opclass tree.Name) error {
+	return pgerror.Newf(pgcode.UndefinedObject, "operator class %q does not exist", opclass)
+}
+
```
Yes, hidden, virtual and stored expressions can be referenced. Updated to add a comment

postamar commented on pkg/sql/schemachanger/scdecomp/decomp.go:
```diff
@@ -571,6 +592,13 @@ func (w *walkCtx) walkIndex(tbl catalog.TableDescriptor, idx catalog.Index) {
+			IndexID: idx.GetID(),
+			Config:  idx.GetGeoConfig(),
+		})
+	}
```
Why is `IndexGeoConfig` its own element instead of the geo config being part of `Index`? It's not something that can be subject to schema changes independently of an index, AFAIK.

postamar commented on pkg/sql/schemachanger/scdecomp/decomp.go:
```diff
@@ -268,6 +268,11 @@ func (w *walkCtx) walkRelation(tbl catalog.TableDescriptor) {
+		w.ev(descriptorStatus(tbl), &scpb.TablePartitionAllBy{
+			TableID: tbl.GetID(),
+		})
+	}
```
This element is not going to age well at all when we actually are going to have to add PARTITION ALL BY support. Consider replacing this with something named `TablePartitioning` which is like `IndexPartitioning` but without the IndexID?

postamar commented on pkg/sql/schemachanger/scdecomp/decomp.go:
```diff
@@ -488,13 +493,29 @@ func (w *walkCtx) walkIndex(tbl catalog.TableDescriptor, idx catalog.Index) {
 				OrdinalInKind: uint32(i),
 				Kind:          scpb.IndexColumn_KEY,
 				Direction:     cpy.KeyColumnDirections[i],
+				Implicit:      implicitColumn,
```
nit: `Implicit: i < idx.ImplicitPartitioningColumnCount()`?

postamar commented on pkg/sql/schemachanger/scpb/elements.proto:
```diff
@@ -72,6 +73,7 @@ message ElementProto {
   RowLevelTTL row_level_ttl = 29 [(gogoproto.customname) = "RowLevelTTL", (gogoproto.moretags) = "parent:\"Table\""];
   TableZoneConfig table_zone_config = 121 [(gogoproto.moretags) = "parent:\"Table, View\""];
   TableData table_data = 131 [(gogoproto.customname) = "TableData", (gogoproto.moretags) = "parent:\"Table, View, Sequence\""];
+  TablePartitionAllBy table_partition_all_by = 132 [(gogoproto.customname) = "TablePartitionAllByl", (gogoproto.moretags) = "parent:\"Table, View, Sequence\""];
```
I don't think views and sequences can be partitioned.

postamar commented on pkg/sql/schemachanger/scdecomp/decomp.go:
```diff
@@ -488,13 +493,29 @@ func (w *walkCtx) walkIndex(tbl catalog.TableDescriptor, idx catalog.Index) {
+				case catpb.InvertedIndexColumnKind_TRIGRAM:
+					invertedKind = scpb.IndexColumn_INVERTED_TRIGRAM
+				}
+			}
```
Why not use the existing catpb enum directly? What purpose does this scpb enum serve?

postamar commented on pkg/sql/schemachanger/scexec/scmutationexec/eventlog.go:
```diff
@@ -110,8 +110,9 @@ func asEventPayload(
 		switch targetStatus {
 		case scpb.Status_PUBLIC:
-			return &eventpb.AlterTable{
+			return &eventpb.CreateIndex{
```
This makes sense in and of itself (and I prefer it) but have you checked what the legacy schema changer does? Making this an eventpb.AlterTable may have been deliberate.

postamar commented on pkg/sql/schemachanger/scexec/scmutationexec/index.go:
```diff
@@ -466,6 +474,14 @@ func (m *visitor) RemoveColumnFromIndex(ctx context.Context, op scop.RemoveColum
+	switch op.InvertedKind {
+	case scpb.IndexColumn_INVERTED:
+		fallthrough
+	case scpb.IndexColumn_INVERTED_TRIGRAM:
```
nit `case scpb.IndexColumn_INVERTED, scpb.IndexColumn_INVERTED_TRIGRAM:` and then you don't need to `fallthrough`

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -131,191 +168,534 @@ func CreateIndex(b BuildCtx, n *tree.CreateIndex) {
+	tempIdxSpec.temporary = &scpb.TemporaryIndex{
+		Index:                    protoutil.Clone(idxSpec.secondary).(*scpb.SecondaryIndex).Index,
+		IsUsingSecondaryEncoding: true,
+	}
```
Do yourself a favour and use `idxSpec.clone()` or, even better, introduce a `makeTempIndexSpec(src indexSpec) (tmp indexSpec)` constructor and use that both there and in `makeSwapIndexSpec`.

postamar reviewed (changes requested):
Thanks for doing this, this is coming together nicely! 

This is only a partial review as I'm requesting changes in the element model which make it pointless to review the data-driven test output. It should be pretty straightforward, though.

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92128)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @msbutler, @postamar, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 550 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqMCKAOSi3Ts4wZ-r4-550:-NHW3UoOE-j8DIks3bT8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fa3e72b4938099107368ee53618741aa8857910/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L550)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

TIL: virtual columns can be used in more than one index!
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 615 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqMCKAOSi3Ts4wZ-r4-615:-NHW3Vk1D7jysCtk_M7I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fa3e72b4938099107368ee53618741aa8857910/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L615)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

similarly, call the existing function to get the column name
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go` line 72 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqV6riqF8K3YcPF-r13-72:-NHW3WRe-xNotD_SkrGX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f73e4040a0a3d71617f82b73b586f1a0f5bf75aa/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go#L72)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

Such a monumental change! Bravo!
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 275 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqRArQN-TPFcjp3-r31-275:-NHW8qZ39PNkGHZ3fm_t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70ef6eeab596d3581aea2e66e64c97a8255d616/pkg/sql/schemachanger/scdecomp/decomp.go#L275)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This element is not going to age well at all when we actually are going to have to add PARTITION ALL BY support. Consider replacing this with something named `TablePartitioning` which is like `IndexPartitioning` but without the IndexID?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 509 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqRArQN-TPFcjp3-r31-509:-NHW9GOH3Tq2tbfzGn8s:b7p071w) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70ef6eeab596d3581aea2e66e64c97a8255d616/pkg/sql/schemachanger/scdecomp/decomp.go#L509)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Why not use the existing catpb enum directly? What purpose does this scpb enum serve?
</blockquote></details>

Good point, this was silly on my part.

Done.
___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 601 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqRArQN-TPFcjp3-r31-601:-NHW9R_lEVglE6rLGvXP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70ef6eeab596d3581aea2e66e64c97a8255d616/pkg/sql/schemachanger/scdecomp/decomp.go#L601)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Why is `IndexGeoConfig` its own element instead of the geo config being part of `Index`? It's not something that can be subject to schema changes independently of an index, AFAIK.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/eventlog.go` line 113 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqTW_6oqDtQfkzW-r31-113:-NHWH6GN8BI4T7o5wIbs:b-jvjest) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70ef6eeab596d3581aea2e66e64c97a8255d616/pkg/sql/schemachanger/scexec/scmutationexec/eventlog.go#L113)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This makes sense in and of itself (and I prefer it) but have you checked what the legacy schema changer does? Making this an eventpb.AlterTable may have been deliberate. 
</blockquote></details>

So, CREATE INDEX in legacy emits eventpb.CreateIndex and ADD COLUMN will emit eventpb.AlterTable. So, I think the question is the right thing in terms of events if someone adds a unique column, we will emit both in the new model, which I think is a reasonable representation.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 99 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqRArQN-TPFcjp7-r9-99:-NHWFDMo1Su3RhXM9w-V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96f7da6c2795b6a4f913d3450ecc8f2089a57be2/pkg/sql/schemachanger/scpb/elements.proto#L99)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

I am curious what is the justification/benefit for introducing new elements for `geo config` and `secondary_index_partial`, instead of adding them as fields in existing elements? 
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 76 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqRArQN-TPFcjp7-r31-76:-NHWFBciBqe_AJXL55AK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70ef6eeab596d3581aea2e66e64c97a8255d616/pkg/sql/schemachanger/scpb/elements.proto#L76)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I don't think views and sequences can be partitioned.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_index_geoconfig.go` line 1 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqSwbcV7yCG6sKG-r8-1:-NHWFEbA9ojvNwBacUrY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e881c871349150bf1ee7ec14e0cc6e9ddba0f2f9/pkg/sql/schemachanger/scplan/internal/opgen/opgen_index_geoconfig.go#L1)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

oh I see it now -- it's for use in the next commit. Please fix it.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/dep_add_index.go:
```diff
@@ -71,21 +71,36 @@ func init() {
+					(*scpb.SecondaryIndex)(nil),
+				),
+				joinOnIndexID(from, to, "table-id", "index-id"),
+				statusesToPublicOrTransient(from, scpb.Status_PUBLIC, to, scpb.Status_WRITE_ONLY),
```
Why is the `to` status not `scpb.Status_VALIDATED` here? Probably nothing bad happens from it being `scpb.Status_WRITE_ONLY` but it's a bit weird.

postamar commented on pkg/sql/schemachanger/scpb/elements.proto:
```diff
@@ -484,6 +488,13 @@ message IndexColumn {
   // Direction is only populated for KEY columns.
   sql.catalog.catpb.IndexColumn.Direction direction = 6;
+
+  // Indicates if this column is implicit.
```
nit: worth noting that this relates to partitioning.

fqazi replied:
Good point:

  // Indicates if this column is implicitly included within the index for
  // partitioning.

postamar commented on pkg/sql/schemachanger/scbuild/testdata/alter_table_add_column:
```diff
@@ -20,7 +20,7 @@ ALTER TABLE defaultdb.foo ADD COLUMN j INT NOT NULL DEFAULT 123
   {columnId: 1, indexId: 1, tableId: 104}
 - [[PrimaryIndex:{DescID: 104, IndexID: 1, ConstraintID: 1}, ABSENT], PUBLIC]
-  {constraintId: 1, indexId: 1, isUnique: true, tableId: 104}
+  {constraintId: 1, geoConfig: {}, indexId: 1, isUnique: true, tableId: 104}
```
If it's not too much trouble, I think it's worth eliding these empty structs from the test output somehow. This could be achieved by embedding the geoConfig struct in the index element, this way `scviz.RewriteEmbeddedIntoParent` would catch it. Perhaps there's a better way, I don't know.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go:
```diff
@@ -491,6 +491,10 @@ func getImplicitSecondaryIndexName(
 		} else {
 			_, _, colName := scpb.FindColumnName(colElts)
+			if idx.Sharding != nil &&
+				colName.Name == idx.Sharding.Name {
```
nit: why the newline after &&?

postamar commented on pkg/sql/schemachanger/scbuild/builder_state.go:
```diff
@@ -517,6 +536,27 @@ func (b *builderState) ComputedColumnExpression(tbl *scpb.Table, d *tree.ColumnT
+	// for single statement transactions.
+	_, err := schemaexpr.ValidatePartialIndexPredicate(
+		b.ctx,
+		b.descCache[tableID].desc.(catalog.TableDescriptor),
```
You should check that `b.descCache[tableID]` exists and that its `desc` can be cast to a TableDescriptor before doing this. Also, above, you need to check that `ns` actually exists (that, in itself, should guarantee the existence of `b.descCache[tableID]`).

postamar commented on pkg/sql/catalog/schemaexpr/partial_index.go:
```diff
@@ -59,7 +59,7 @@ func ValidatePartialIndexPredicate(
 		return "", err
 	}
-	if !desc.IsNew() {
+	if mutDesc, ok := desc.(catalog.MutableTableDescriptor); !ok || !mutDesc.IsNew() {
```
Perhaps checking that the version > 1 would be equivalent? Just a thought.

fqazi replied:
I don't think it would be equivalent since it doesn't guarantee that the object was created in the current transaction. i.e. It's valid to modify the first version of the object

postamar replied:
Good point!

postamar commented on pkg/ccl/schemachangerccl/testdata/explain/drop_table_multiregion:
```diff
@@ -48,9 +49,10 @@ Schema change plan for DROP TABLE ‹multi_region_test_db›.‹public›.‹tab
+ │         └── 16 Mutation operations
  │              ├── MarkDescriptorAsDropped {"DescriptorID":108}
  │              ├── RemoveAllTableComments {"TableID":108}
+ │              ├── NotImplemented {"ElementType":"scpb.TablePartit..."}
```
Ruh roh. Looks like you need to add `TablePartitioning` to that set of elements which will be no-opped when dropping tables and the like. The op-rule in question is named `skip element removal ops on descriptor drop`.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -103,20 +140,20 @@ func CreateIndex(b BuildCtx, n *tree.CreateIndex) {
 			// dropped columns.
 			if target == scpb.ToPublic {
-				source = t
+				idxSpec.primary = t
```
This is strange. In my view `idxSpec` only pertains to this new secondary index and `primary` should never be set. Besides, you set it to `nil` soon after.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -103,20 +140,20 @@ func CreateIndex(b BuildCtx, n *tree.CreateIndex) {
 	})
 	if n.Unique {
-		index.ConstraintID = b.NextTableConstraintID(index.TableID)
+		idxSpec.secondary.ConstraintID = b.NextTableConstraintID(idxSpec.secondary.TableID)
```
Shouldn't we be setting `idxSpec.secondary.IsUnique = true` here as well? I found out that we set it above when setting `idxSpec.secondary` but I think it would make more sense here.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -131,191 +168,518 @@ func CreateIndex(b BuildCtx, n *tree.CreateIndex) {
+		} else if ic.Kind == scpb.IndexColumn_KEY_SUFFIX {
+			ic.OrdinalInKind = uint32(keySuffixIdx)
+			keySuffixIdx++
+		}
```
What about STORED columns?

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:
```diff
@@ -131,191 +168,518 @@ func CreateIndex(b BuildCtx, n *tree.CreateIndex) {
+		TableID: idxSpec.secondary.TableID,
+		IndexID: idxSpec.secondary.IndexID,
+		Name:    indexName,
+	})
```
Since you're using `idxSpec` don't you want to set the name there instead?

fqazi replied:
I'll add a comment to clarify the rationale here, but we can't compute implicit names without adding elements. I think we could avoid similar funniness if we had an element result from an array or something or with synthetic things.

	// Apply the name once everything else has been created since we need
	// elements to be added, so that getImplicitSecondaryIndexName can make
	// an implicit name if one is required.

postamar reviewed (commented):
Looks like my comment about `typeCheckConstant` wasn't addressed, can you follow up on that please? See also my other comment about `idxSpec.clone()`.

In any case thanks for the quick turnaround. I've looked at the rest of the PR and here are my additional comments. I look forward to approving this! Nicely done.

fqazi commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_secondary_index.go:
```diff
@@ -86,6 +86,14 @@ func init() {
+				emit(func(this *scpb.SecondaryIndex) *scop.RefreshStats {
+					return &scop.RefreshStats{
+						TableID: this.TableID,
+					}
```
That's a good point, let me get an issue open this is very blanket in the legacy schema changer. In the declarative one we can be more deliberate about which operations should do this.

fqazi committed:
sql: force types to be hydrated when evaluating constants

Previously, there were scenarios where the declarative schema changer
validating expressions could run into cases where types were not fully
evaluated. This patch addresses one such case by force fully hydrating
a type when parsing/evaluating constants in expressions.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92128)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @msbutler, @postamar, and @Xiang-Gu)

___
*[`pkg/sql/catalog/schemaexpr/partial_index.go` line 62 at r43](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqSwbcV7yCG6sKF-r43-62:-NHeitUb9otvVLPGVLxT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cf3263c00c33ef9472c187b2152d2e5a098af85/pkg/sql/catalog/schemaexpr/partial_index.go#L62)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Good point!
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 550 at r43](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqMCKAOSi3Ts4wY-r43-550:-NHeiuUwAvKSfSotTp27:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cf3263c00c33ef9472c187b2152d2e5a098af85/pkg/sql/schemachanger/scbuild/builder_state.go#L550)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

You should check that `b.descCache[tableID]` exists and that its `desc` can be cast to a TableDescriptor before doing this. Also, above, you need to check that `ns` actually exists (that, in itself, should guarantee the existence of `b.descCache[tableID]`). 
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 224 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqMCKAOSi3Ts4wZ-r31-224:-NHeixtd1l4ojMlx4SEt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70ef6eeab596d3581aea2e66e64c97a8255d616/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L224)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Do yourself a favour and use `idxSpec.clone()` or, even better, introduce a `makeTempIndexSpec(src indexSpec) (tmp indexSpec)` constructor and use that both there and in `makeSwapIndexSpec`. 
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 143 at r43](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqMCKAOSi3Ts4wZ-r43-143:-NHeivop9LNLRkpvX7NW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cf3263c00c33ef9472c187b2152d2e5a098af85/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L143)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This is strange. In my view `idxSpec` only pertains to this new secondary index and `primary` should never be set. Besides, you set it to `nil` soon after.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 148 at r43](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqMCKAOSi3Ts4wZ-r43-148:-NHeiwSRDK0txpLDynQg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cf3263c00c33ef9472c187b2152d2e5a098af85/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L148)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't we be setting `idxSpec.secondary.IsUnique = true` here as well? I found out that we set it above when setting `idxSpec.secondary` but I think it would make more sense here.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 194 at r43](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqMCKAOSi3Ts4wZ-r43-194:-NHeixC6EOTWkLh34Ik8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cf3263c00c33ef9472c187b2152d2e5a098af85/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L194)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

What about STORED columns?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_add_index.go` line 100 at r43](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAVoqTW_6oqDtQfkzY-r43-100:-NHeiz7-1SHJHs4Wxl7P:bibwxk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cf3263c00c33ef9472c187b2152d2e5a098af85/pkg/sql/schemachanger/scplan/internal/rules/dep_add_index.go#L100)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Why is the `to` status not `scpb.Status_VALIDATED` here? Probably nothing bad happens from it being `scpb.Status_WRITE_ONLY` but it's a bit weird.
</blockquote></details>

Yes, good point I changed this to validated.

Done.
___
*[`pkg/sql/sem/tree/constant.go` line 79 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHCi_-MakLcUl3zl0v8-r15-79:-NHexJ4qBmiy3O5Qwocj:bm1d6ir) ([raw file](https://github.com/cockroachdb/cockroach/blob/fecb571af812c110dd7e79ef674bc90a20c1f9df/pkg/sql/sem/tree/constant.go#L79)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'm highly suspicious of this change. For a start, there's no guarantee that semaCtx.TypeResolver is not nil. Also, why isn't this new logic pushed down into the corresponding ResolveAsType implementation instead of being added here?
</blockquote></details>

I moved this into the ResolveAsType implementation for enums where this issue was being observed. Additionally, added an assertion for confirming that a resolver is there if hydration needs to be done here.
___
*[`pkg/ccl/schemachangerccl/testdata/explain/drop_table_multiregion` line 55 at r43](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHAi7Hupr5gteZrVxDv-r43-55:-NHenUxXDha_KuutWS_m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cf3263c00c33ef9472c187b2152d2e5a098af85/pkg/ccl/schemachangerccl/testdata/explain/drop_table_multiregion#L55)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Ruh roh. Looks like you need to add `TablePartitioning` to that set of elements which will be no-opped when dropping tables and the like. The op-rule in question is named `skip element removal ops on descriptor drop`.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/testdata/alter_table_add_column` line 23 at r43](https://reviewable.io/reviews/cockroachdb/cockroach/92128#-NHWUxcMdRX_6fxI6ZZn-r43-23:-NHenVovBCnnL9ozDE29:bcjv369) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cf3263c00c33ef9472c187b2152d2e5a098af85/pkg/sql/schemachanger/scbuild/testdata/alter_table_add_column#L23)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

If it's not too much trouble, I think it's worth eliding these empty structs from the test output somehow. This could be achieved by embedding the geoConfig struct in the index element, this way `scviz.RewriteEmbeddedIntoParent` would catch it. Perhaps there's a better way, I don't know.
</blockquote></details>

I changed this to a nullable value, which has the same effect. Embedding this structure has the problem where it has a super vague name called Config, so I'm afraid of the unambiguous name.


<!-- Sent from Reviewable.io -->

fqazi commented:
TFTR @postamar @ajwerner @Xiang-Gu

fqazi commented:
bors r+

fqazi commented:
bors r-

fqazi commented:
bors r+

fqazi commented:
bors r+

fqazi commented:
bors r-

fqazi commented:
bors r+

fqazi commented:
bors r+

fqazi commented:
bors r+

fqazi commented:
The only failure observed is a known flake inside:  TestEvictionTokenCoalesce.

bors r+
